### PR TITLE
Support CLI-specific import file statements

### PR DESCRIPTION
**Pull request by stefansperber**
I added support for `source` (HiveCli) and `!run` (Beeline) commands. These commands allow to import and execute scripts in statements or scripts. This allows to compose and reuse HQL scripts.

Since these commands are not part of the core language I limited changes to the parts that emulate CLI-specific behaviour. Let me know what you think.

Some remarks:

(1) To keep `HiveServerContainer` agnostic to the used CLI I moved splitting scripts into statements out of the class. Afterwards, I removed the now unused `HiveServerContainer#executeScript` (it does not support `source/!run`). This breaks backwards compatibility. If deprecation is a better option commit `2d495443b8cc31ac99ca7c7fec77e4b193d78f0b` can be reverted.

(2) I noticed that [Beeline's SQLLine](https://cwiki.apache.org/confluence/display/Hive/HiveServer2+Clients#HiveServer2Clients-BeelineCommands) commands are not supported/emulated (for example, `!run` or `!quit`). If these should be supported in the future further refactorings could be done.

Here's an idea instead of string statements:
```
class Statement {

  private final StatementType type;
  private final String value;
  
  // ... constructor + getters for members omitted
  
  String[] getTokens() {
    return value.trim().split("\\s+");
  } 
}
```
```
enum StatementType {
  /* */
  HQL,
  /* supported by both CLIs */
  IMPORT_FILE,
  /* */
  HIVE_CLI_SPECIFIC_COMMAND,
  /* */
  BEELINE_SPECIFIC_COMMAND,
  
  // ...
}
```

This would also simplify my changes a bit more.

Example usage:
```
for (Statement statement : StatementsSplitter.splitStatements(commandShellEmulation.transformScript(script))) {
  if (StatementType.HQL == statement.getType()) {
    execute(statement);
  } else if (StatementType.IMPORT_FILE == statement.getType()) {
  	// ...
  }
}
```

### CODE REVIEW

**Comment by stefansperber** [comment]
Test failures do not seem to be related to changes in the branch.

File: src/main/java/com/klarna/hiverunner/builder/HiveShellBase.java [hunk 1]
```diff
@@ -100,7 +102,21 @@
...
+      return hiveServerContainer.executeStatement(transformedHql);
+    }
+
+    private List<Object[]> importScript(String hql) {
+      String[] tokens = hql.trim().split("\\s+");
+      String fileName = null;
+      if (tokens.length == 2) {
+        fileName = tokens[1];
+      }
+      execute(new File(fileName));
```
Comment (massdosage) [inline src/main/java/com/klarna/hiverunner/builder/HiveShellBase.java]: Do you want to do this if fileName is null? Probably not? Maybe move this into the above if statement. Is it OK to fail silently (i.e. do nothing) if fileName is null? When would this happen?

File: src/test/java/com/klarna/hiverunner/sql/StatementsSplitterTest.java [hunk 2]
```diff
@@ -160,4 +161,16 @@ public void testReadUntilEndOfLine() {
...
+        String statementB = "select * from table where foo != bar";
+        String statementC = "!run another_script.hql";
+
+        List<String> expected = Arrays.asList(statementA, statementB, statementC);
+        String expression = statementA + '\n' + statementB + ";   " + statementC;
+
+        Assert.assertEquals(expected, StatementsSplitter.splitStatements(expression));
+    }
+
 }
```
Comment (massdosage) [inline src/test/java/com/klarna/hiverunner/sql/StatementsSplitterTest.java]: You should also have a test for empty statements and to check the trimming you do in the class under test.

File: src/test/java/com/klarna/hiverunner/builder/HiveShellBaseTest.java [hunk 3]
```diff
@@ -148,6 +151,42 @@ public void executeScriptNotStarted() throws IOException {
...
+      String importHql = "source " + importedFile.getAbsolutePath();
+      File file = new File(tempFolder.getRoot(), "script.hql");
+      Files.write(importHql, file, UTF_8);
+
+      shell.execute(file);
+
+      verify(container).executeStatement(hql);
+    }
+
     private HiveShell createHiveShell(String... keyValues) {
```
Comment (massdosage) [inline src/test/java/com/klarna/hiverunner/builder/HiveShellBaseTest.java]: What about a test for whatever would result in a null file path ended up in the class under test?

File: src/main/java/com/klarna/hiverunner/builder/HiveShellBase.java (hunk 1 above)
Comment (stefansperber) [inline src/main/java/com/klarna/hiverunner/builder/HiveShellBase.java]: I can see `fileName` be null when the script contains and error. For example, `source;`. In this case, a `NullPointerException` will be thrown ([File Javadoc](https://docs.oracle.com/javase/7/docs/api/java/io/File.html#File(java.lang.String))).

We could throw an `IllegalArgumentException` instead. This way it's consistent with the behaviour in `HiveShellBuilder#setScriptsUnderTest` when a script file cannot be loaded.

File: src/test/java/com/klarna/hiverunner/sql/StatementsSplitterTest.java (hunk 2 above)
Comment (stefansperber) [inline src/test/java/com/klarna/hiverunner/sql/StatementsSplitterTest.java]: Good point. Will add one.

File: src/test/java/com/klarna/hiverunner/builder/HiveShellBaseTest.java (hunk 3 above)
Comment (stefansperber) [inline src/test/java/com/klarna/hiverunner/builder/HiveShellBaseTest.java]: Another good point. Will add one to document the expected behaviour.

File: src/main/java/com/klarna/hiverunner/HiveServerContainer.java [hunk 4]
```diff
@@ -132,16 +132,6 @@ public TemporaryFolder getBaseDir() {
     }
 
     /**
-     * Executes a hive script.
-     * @param hiveql hive script statements.
-     */
-    public void executeScript(String hiveql) {
```
Comment (teabot) [inline src/main/java/com/klarna/hiverunner/HiveServerContainer.java]: I'm unclear why this API breaking change is needed. Surely we must retain the distinction between a script and a statement?

File: src/main/java/com/klarna/hiverunner/sql/StatementsSplitter.java [hunk 5]
```diff
@@ -33,13 +33,15 @@
...
+    public static final String BEELINE_SPECIAL_CHARS = "!";
 
     /**
      * Splits expression on ';'.
+     * Also, detects Beeline's SQLLine commands (commands start with '!' and do not end with ';').
      * ';' within quotes (" or ') or comments ( -- ) are ignored.
      */
     public static List<String> splitStatements(String expression) {
-        StringTokenizer tokenizer = new StringTokenizer(expression, SQL_SPECIAL_CHARS, true);
+        StringTokenizer tokenizer = new StringTokenizer(expression, SQL_SPECIAL_CHARS + BEELINE_SPECIAL_CHARS, true);
```
Comment (teabot) [inline src/main/java/com/klarna/hiverunner/sql/StatementsSplitter.java]: Beeline has leaked into what should be a general statement splitter class. Is there anyway this can be avoided?

File: src/test/java/com/klarna/hiverunner/builder/HiveShellBaseTest.java [hunk 6]
```diff
@@ -148,6 +151,42 @@ public void executeScriptNotStarted() throws IOException {
...
+      HiveShell shell = createHiveShell();
+      shell.start();
+
+      String importHql = "source " + importedFile.getAbsolutePath();
+      File file = new File(tempFolder.getRoot(), "script.hql");
+      Files.write(importHql, file, UTF_8);
+
+      shell.execute(file);
+
+      verify(container).executeStatement(hql);
```
Comment (teabot) [inline src/test/java/com/klarna/hiverunner/builder/HiveShellBaseTest.java]: I assume for a multi-line statement in `hql` we'd see multiple invocations of `container.executeStatement(...)`?

**Comment by PelleUllberg** [comment]
Hi, 

Sorry for the inactivity but I've been on one of those nice long Swedish vacations again. I'm a bit choked at the moment, but will try to review this pull request in a week or two. 

Awesome feeling to get back to work and find new committers for HiveRunner!

Also, thanks a ton to @teabot for the hours you put into this project!

regards
/Pelle

**Comment by Thoymas** [comment]
@stefansperber Could you merge up the latest changes from hiverunner master so we can see if Travis can build your pull-request?

**Comment by stefansperber** [comment]
@Thoymas Merged your changes from `master`. Thanks for looking into the CI issue.

**Comment by stefansperber** [comment]
I had another look at how `BeeLine` and `CliDriver` handle their import statements, and I found that my first implementation was making wrong assumptions. So I moved the logic for handling `source`/`!run` commands into `CommandShellEmulation`.

- The Hive emulation closely mirrors what `CliDriver` does (also, in the case of wrong command syntax). [Source](https://github.com/apache/hive/blob/0ae374a320d1cae523ba2b434800e97692507db8/cli/src/java/org/apache/hadoop/hive/cli/CliDriver.java#L137)

- Emulation for Beeline was less straightforward as `Commands#run` does not throw an exception when something goes wrong. Instead, a `false` value is returned which is later used to make the exit code. [Source](https://github.com/apache/hive/blob/master/beeline/src/java/org/apache/hive/beeline/Commands.java#L1535)

I chose to throw an `IllegalArgumentException` when we cannot get the filename in case the import statement's syntax is wrong. Not sure if this is the best solution; maybe both shell emulations should have similar error handling.

**Comment by stefansperber** [comment]
@teabot 

(1) `StatementsSplitter`

Like you said. It's not great that Beeline gets special treatment in the generic `StatementsSplitter`.

I'm struggling to come up with a better alternative. Extract a `StatementSplitter` interface, and split the class up into base class and two shell-specific implementations? Or have a hook method that can be overridden to add Beeline-specific behaviour? Hope you have an idea. Please advise :)

(2) Changelog: Updated.

File: src/test/java/com/klarna/hiverunner/builder/HiveShellBaseTest.java (hunk 6 above)
Comment (stefansperber) [inline src/test/java/com/klarna/hiverunner/builder/HiveShellBaseTest.java]: `HiveShell` will read the file into a string, which is then executed the same way as a multi-statement script. Do you think this needs a extra testing? I'm reusing existing functionality.

File: src/main/java/com/klarna/hiverunner/sql/StatementsSplitter.java (hunk 5 above)
Comment (stefansperber) [inline src/main/java/com/klarna/hiverunner/sql/StatementsSplitter.java]: Added a comment for this on the pull request discussion page.

File: src/main/java/com/klarna/hiverunner/HiveServerContainer.java (hunk 4 above)
Comment (stefansperber) [inline src/main/java/com/klarna/hiverunner/HiveServerContainer.java]: Import statements can be used in statements and scripts. Both `CliDriver` and `BeeLine` will first read a script file in, split its lines into statements and then execute them individually. This way they can support their custom commands.

File: src/test/java/com/klarna/hiverunner/builder/HiveShellBaseTest.java (hunk 3 above)
Comment (stefansperber) [inline src/test/java/com/klarna/hiverunner/builder/HiveShellBaseTest.java]: My commit from yesterday and today moved the logic for getting the file to import into `CommandShellEmulation`. I've added tests for valid and malformed `source`/`!run` commands there. I'm waiting for a verdict on these changes; can then add extra tests for when there's trouble with the script file in `HiveShellBase`.

File: src/test/java/com/klarna/hiverunner/sql/StatementsSplitterTest.java (hunk 2 above)
Comment (stefansperber) [inline src/test/java/com/klarna/hiverunner/sql/StatementsSplitterTest.java]: I noticed that `expression` already adds whitespace before `statementC` to verify trimming. Do you think it's OK?

File: README.md [hunk 7]
```diff
@@ -227,7 +227,11 @@ Future work and Limitations
 Change Log (From version 2.2.0 and onwards)
 ==============
 
-### __2.6.0-SNAPSHOT__
+### __2.6.1-SNAPSHOT__
```
Comment (PelleUllberg) [inline README.md]: Minor version should be bumped since we added functionality -> 2.7.0
Also remember to bump the version in the pom.xml file to 2.7.0

File: README.md [hunk 8]
```diff
@@ -227,7 +227,11 @@ Future work and Limitations
...
+
+Support shell-specific `source` (`hive`) and ``!run`` (`beeline`) commands. (These commands allow to import and execute scripts in statements or scripts.)
```
Comment (PelleUllberg) [inline README.md]: I think that this maybe deserves it's own section in the README

File: src/main/java/com/klarna/hiverunner/CommandShellEmulation.java [hunk 9]
```diff
@@ -35,6 +53,25 @@ public String transformScript(String script) {
...
+      String[] tokens = statement.trim().split(" ");
+      if (tokens.length == 2) {
+        return new File(tokens[1]);
+      }
+      throw new IllegalArgumentException("Cannot get file to import from '" + statement + "'");
+    }
+
     @Override
     public String transformStatement(String statement) {
       return filterFullLineComments(statement);
```
Comment (PelleUllberg) [inline src/main/java/com/klarna/hiverunner/CommandShellEmulation.java]: According to this http://sqlline.sourceforge.net/#sect_command_run any rows starting with '#' should be ignored

File: src/main/java/com/klarna/hiverunner/builder/HiveShellBase.java [hunk 10]
```diff
@@ -100,7 +102,16 @@
     }
```
Comment (PelleUllberg) [inline src/main/java/com/klarna/hiverunner/builder/HiveShellBase.java]: Why can't all command shell specific transformation happen in the transformScript call. The returned sql would have all the imports/!run commands etc. exploded and ready to be executed by the regular hive parser. This way it's easy to add more of the beeline commands etc without leaking outside of the emulator

File: src/main/java/com/klarna/hiverunner/HiveServerContainer.java (hunk 4 above)
Comment (teabot) [inline src/main/java/com/klarna/hiverunner/HiveServerContainer.java]: Ok, so am I correct in thinking that fundamentally Hive only executes statements and that the concept of a script is implemented at the Beeline/CLI layer? If so then I can see why it'd make sense to remove `executeScript`.

File: src/main/java/com/klarna/hiverunner/sql/StatementsSplitter.java (hunk 5 above)
Comment (teabot) [inline src/main/java/com/klarna/hiverunner/sql/StatementsSplitter.java]: Perhaps this could be an instance belonging to {{BEELINE}} and {{HIVE_CLI}} that is instantiated with the sets of special characters via a default constructor?

**Comment by PelleUllberg** [comment]
@stefansperber Any update on this PR?

**Comment by stefansperber** [comment]
@PelleUllberg and everybody else: Thank you very much for taking your time to give me feedback. Apologies for the long silence! I should have replied earlier.

I'll start working on making the changes this week. Was a bit swamped with work recently.

File: src/main/java/com/klarna/hiverunner/CommandShellEmulation.java (hunk 9 above)
Comment (stefansperber) [inline src/main/java/com/klarna/hiverunner/CommandShellEmulation.java]: The implementation seems to only support `!<command>` (see [BeeLine#dispatch](https://github.com/apache/hive/blob/master/beeline/src/java/org/apache/hive/beeline/BeeLine.java#L1128)).

File: src/main/java/com/klarna/hiverunner/HiveServerContainer.java (hunk 4 above)
Comment (stefansperber) [inline src/main/java/com/klarna/hiverunner/HiveServerContainer.java]: From my understanding that is how Beeline and Hive CLI do it.

**Comment by stefansperber** [comment]
I've pushed some new changes. They are just sketches and I hope they can show where I'd like to take the implementation. Please ignore the individual commits; names and locations should also be improved.

When I went through the feedback from before I saw these issues:

1. I like the idea of including all scripts before we execute the statements. This leads to a much better separation of shell-specific behaviour. I don't want to duplicate the code for reading scripts from files/paths from `HiveShellBase` tho.

2. Charset support. The `HiveShellBase#execute` methods for files and paths also accept other `Charset`s. This seems difficult to support when we're handling single import statements.

Moving forward, I think it's best to close this pull request and submit two new ones:

- Refactor `HiveShell` and drop the methods that take other `Charset`s. Instead, set a (non-default) charset when `HiveShellBase` is constructed. This charset is used for reading all files. `HiveRunner` version should be bumped to `3.0.0`.

- Introduce `HiveQueryLanguageStatement` and move all code that deals with files/paths, include statements, ... into a factory. `HiveShellBase` does not know about splitting and transforming statements anymore but only handles collecting results and calling `HiveServerContainer`.

What do you think?

File: src/main/java/com/klarna/hiverunner/hql/HiveQueryLanguageStatementFactory.java [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+		for (String statement : statements) {
+			hqlStatements.addAll(newInstanceForStatement(statement));
+		}
+		return hqlStatements;
+	}
+
+	public List<HiveQueryLanguageStatement> newInstanceForStatement(String statement) {
+		List<HiveQueryLanguageStatement> hqlStatements = new ArrayList<>();
+		String trimmedStatement = statement.trim();
+		if (commandShellEmulation.isImportFileStatement(trimmedStatement)) {
```
Comment (PelleUllberg) [inline src/main/java/com/klarna/hiverunner/hql/HiveQueryLanguageStatementFactory.java]: I still think this will become very bloated if more commands are implemented.
There should be a clean separation between implementations of the different shells and the specific commands. I rather we pass the collection of statements through the emulator and it manipulates it in whatever way needed to simulate the commands.

**Comment by stefansperber** [comment]
Closing this for now. Unfortunately, I cannot find the time that this PR deserves at the moment. Apologies for that. Hope somebody else can base their changes on the discussion that we had and will succeed in adding this useful feature.
